### PR TITLE
[IA-3765] Add V2 workspace based app routes

### PR DESCRIPTION
**Pull request by n8landolt**
https://broadworkbench.atlassian.net/browse/IA-3765
---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green (you can re-run automation tests with a comment saying `jenkins retest`
- [ ] Run the automation tests multiple times in parallel to weed out instability if applicable via a comment saying `multi-test`
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: http/src/main/resources/swagger/api-docs.yaml
Comment (Qi77Qi) [inline http/src/main/resources/swagger/api-docs.yaml]: add doc for all apis

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: add TODO here

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala]: is the comment accurate? think we can remove `machineType` for now

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: this permission part should be mostly the same as v1 implementation. might make sense to have a helper function to share the code

File: http/src/main/resources/swagger/api-docs.yaml
Comment (rtitle) [inline http/src/main/resources/swagger/api-docs.yaml]: Should we add the other routes (getApp, createApp, deleteApp) to swagger as well?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala:527]: `WorkspaceAction.CreateControlledUserResource` seems like the right thing to check here 👍 (even though we're not actually making a controlled resource ;))

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: It should, I think.. the commented code here looks mostly correct (we just don't want to create apps with a `googleProject` parent resource in this case I think).

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: Yup we'll need to get a pet managed identity for the azure case.. it's a separate Sam call: https://sam.dsde-dev.broadinstitute.org/#/Azure

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: I think short term we can just not support diskConfig for Azure apps. (We should continue to support it for GCP though.)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala:623]: We should confirm `getSavableApp` populates the right helm chart on both clouds. On GCP we should continue to support our existing charts. For Azure we should only support the `Cromwell` appType and use the chart defined here: https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/reference.conf#L209

Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala:623]: Thinking out loud: should we have a new `AppType` enum for Cromwell-on-Azure?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: Great to see PubSub support started! (even if TODO :)).

Eventually the azure pubsub handling should invoke methods on https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AKSAlgebra.scala

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: Not sure this is the case -- even for Azure apps Leo is creating the Sam resource I think (so we need to delete it)

**Review by rtitle — COMMENTED** [review]
Made some comments (hope you don't mind) but looking good so far! We can always iterate on TODOs in separate PRs too.

File: http/src/main/resources/swagger/api-docs.yaml
Comment (n8landolt) [reply http/src/main/resources/swagger/api-docs.yaml]: Added!

File: http/src/main/resources/swagger/api-docs.yaml
Comment (n8landolt) [reply http/src/main/resources/swagger/api-docs.yaml]: Done! Could use a once over from another set of eyes

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (n8landolt) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: Should I create a more generic version of this in this ticket?

Comment (n8landolt) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: Could you review my updated implementation?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (n8landolt) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: Could you review my updated implementation?

**Comment by rtitle** [comment]
I see the new changes, will look again tomorrow. The Sam stuff is tricky, want to think about it a bit

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala]: It looks like `creatorEmail` is not used at all in this method and `cloudContext` is just used for logging. Thinking maybe we log `"${sr.resourceType.asString}/${sr.resourceIdAsString(resource)}"` instead and just remove these params?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala:187]: Noting there is a slight change compared to `createResource`. In `createResource` we request the pet key based on the `creatorEmail`, and make the Sam calls as the pet. In this method we call Sam with the user token directly.

I think using the token directly is probably fine for `createResourceV2`, since it comes from Front Leo (where we always have the token). However we might need to adjust this when we call `deleteResourceV2` from Back Leo (where we don't have the user token).

I think we can probably start with this though, and revisit as part of [adding deleteApp](https://broadworkbench.atlassian.net/browse/TOAZ-230) (I think there are a couple workable solutions..)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala]: I don't think this is right -- it's using `Project` as the resource type and the workspace id. I think the best thing to do for now is continue to use `SamResourceType.Project` and `ProjectSamResourceId(googleProject)` if the cloudContext is GCP -- and just not set a parent if the cloudContext is Azure. 

(We lose the inherited project-owner permissions in that case -- but honestly we'll be revisiting the Sam stuff anyway when we implement multi-tenancy.)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala]: Re my other comment, maybe we only call `createResourceWithParentV2` if the resourceType is App _and_ the cloud is GCP. Otherwise just `createResourceV2`.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala]: Same as above: the `creatorEmail` and `cloudContext` aren't really used here.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala:187]: Also, is the only difference between `createResource` and `createResourceV2` the fact that the former uses a GCP pet to make the call?

Maybe we could name the `SamDAO` methods based on that difference in behavior, instead of just V1/V2:

```
def createResourceAsGcpPet(resource: R, creatorEmail: WorkbenchEmail, googleProject: GoogleProject)

def createResourceAsUser(resource: R, userInfo: UserInfo)

// etc
```

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala]: Maybe change the metric name

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: Yeah we call `notifyResourceDeletedV2` here, but only if the status is `Error`.. otherwise the Sam deletion will happen in Back Leo after polling for deletion is done. [example for GKE](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala#L1196)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: Looks good

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala
Comment (n8landolt) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala:187]: I'll keep the implementation as is!

I like the name adjustments, I'll do that.

File: http/src/main/resources/swagger/api-docs.yaml
Comment (rtitle) [inline http/src/main/resources/swagger/api-docs.yaml]: Minor nit -- maybe change this text to
```
Apps API v2. Support for Google Kubernetes Engine and Azure Kubernetes Service.
```

(since the apps stuff is all kubernetes.)

Also technically v2 create/delete isn't supported as of this PR (the pubsub messages don't do anything yet ;)). I think it's still fine to add the routes to Swagger, maybe just add `Under active developement.` or something.

**Review by rtitle — APPROVED** [review]
Thumbing this! There may be some iteration/bugfixes once we wire things together but I think this is a good first pass at the route support.

Please make sure tests pass before merge -- would probably be good to get @Qi77Qi thumb too.

File: http/src/main/resources/swagger/api-docs.yaml
Comment (Qi77Qi) [reply http/src/main/resources/swagger/api-docs.yaml]: yea, apps are only on k8s

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (n8landolt) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: What do we do about the NodepoolLeoID?
seems like we're missing machineConfig and clusterId from the code block in createApp.

Comment (n8landolt) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: @Qi77Qi @rtitle

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala]: maybe `deleteResourceWithUserInfo` instead of `deleteResourceV2`? `deleteResourceV2` reads to me we're using v2 api or sth

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala:545]: so we did the TODO?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: where will disk be coming from for say cromwell chart without it?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: we can leave this TODO as is for now...it'll come from WSM api which doesn't support it yet

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: Add TODO


We should add TODO in all places where we're populating temporary values. Otherwise, it'll be pretty hard to find these spots once the PR is merged...For larger items, tickets will be ideal

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala:663]: as I mentioned in slack, if these are all just copy from v1 implementation, let's use a common function

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (n8landolt) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala:623]: I'm not entirely sure how to address this portion.

Do I add another AppType, create CromwellOnAzureAppConfig with the same fields referenced in your link and add a new case under getSavableApp?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala
Comment (n8landolt) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala]: I like it - I'll actually adjust the createResourceAsUser to be WithUserInfo for consistency.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (n8landolt) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: I have another commit incoming, the TODO will be moved to a different spot.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: Cromwell on Azure chart doesn't have a persistent disk parameter -- it mounts a postgres DB and writes logs/etc to a blob storage container. So this implementation is fine I think..

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala:623]: What is the resulting data in the DB after `getSavableApp` is called, on both clouds? (Maybe this would be a good unit test).

In particular this logic might need to be changed: https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala#L687-L691

Currently it handles 3 app types: Galaxy, Cromwell, and Custom. Cromwell-on-Azure is a separate helm chart though, so there should be some case which results in populating this information from conf: https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/reference.conf#L209-L216

I wasn't sure if we want to make `CromwellOnAzure` a separate AppType, or have logic like:
```
appType match {
  case Galaxy   => config.leoKubernetesConfig.galaxyAppConfig
  case Cromwell if cloud == GCP => config.leoKubernetesConfig.cromwellAppConfig
  case Cromwell if cloud == Azure => config.leoKubernetesConfig.coaAppConfig
  case Custom   => config.leoKubernetesConfig.customAppConfig
}
```

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala]: Yeah for Azure we don't actually care about the nodepool and cluster in the Leo DB, since all that comes from the landing zone.

**Review by Qi77Qi — APPROVED** [review]
wanna give a thumb before I go on vacation....

the only thing we need to be careful is not to break existing app APIs ideally. Once tests all pass, things should be mostly okay. But keep an eye on any reported issues on app APIs from @yonghaoy

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/kubernetesModels.scala
Comment (rtitle) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/kubernetesModels.scala]: The logic here is kind of tricky actually. :) 

Thinking through the cases, I think we want to generate v1 or v2 style URLs based on the API being called. But we probably shouldn't fail if there are apps in the system that we can't generate a URL for -- e.g. if I call `listAppsV2` and I have some v1 apps in the system.

I came up with this (warning: didn't compile!). Does this cover all the cases? Btw I like using a pattern match to describe this type of thing.

```
def getProxyUrls(cloudContext: CloudContext,
                   workspaceId: Option[WorkspaceId],
                   proxyUrlBase: String,
                   apiVersion: String
): Map[ServiceName, URL] =
  appResources.services.flatMap { service =>
    val proxyPathOpt = (apiVersion, cloudContext, workspaceId) match {
      // v1 apis should always generate urls for google apps
      case ("v1", CloudContext.Gcp(project), _) =>  
        Some(s"google/v1/apps/${project.value}/${appName.value}/${kubernetesService.config.name.value}")
      // v2 apis should generate urls for apps on both clouds, but only if there's a workspace
      case ("v2", _, Some(workspace)) => 
        Some(s"apps/v2/${workspaceId.value}/${appName.value}/${kubernetesService.config.name.value}")
      // otherwise don't generate a proxy url
      case _ =>
        None
    }
    proxyPathOpt match {
      case Some(proxyPath) =>
        Map(kubernetesService.config.name -> new URL(s"${proxyUrlBase}${proxyPath}${kubernetesService.config.path.getOrElse("")}"))
      case None =>
        Map.empty
    }
  }
```

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/appRoutesModels.scala
Comment (rtitle) [inline core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/appRoutesModels.scala]: Can we remove this TODO comment?

**Review by rtitle — COMMENTED** [review]
One suggestion above w.r.t. proxyUrl generation.

Overall this is looking good. There may be some things to fix once we start wiring the v2 routes into Back Leo, but I think any issues can be fixed in subsequent PRs. Please just make sure these are no regressions in the v1 app routes (I think there are swatomation tests to cover these?), then I think this is probably mergeable.

**Comment by yonghaoy** [comment]
Hello as AoU starts integrating with Leo APP apis, few questions I have: 

- Does this support GCP workspace? If so, does it support Rawls powered GCP workspace or WSM powered workspace? 
-  Is that something AoU can/should start using?

**Comment by rtitle** [comment]
@yonghaoy 

The end state is the v2 app routes will support GCP and Azure. But the v2 routes do require a WSM-powered workspace. Since only Azure workspaces are WSM-powered at the moment, I think that means in the short term that clients should continue using v1 for GCP and v2 for Azure.

File: core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/kubernetesModels.scala
Comment (n8landolt) [reply core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/kubernetesModels.scala]: This looks great! I fixed up the errors and I like it a lot more than my first solution.
Mostly did it as a first pass, so I appreciate the follow up.

**Review by rtitle — APPROVED** [review]

**Comment by rtitle** [comment]
jenkins retest

**Pull request merged**
